Don’t pass Traversable as ArrayUtils::merge parameter

gszy opened the pull request:
`SessionManager::start()` retrieves data from `$_SESSION` and passes it to `ArrayUtils::merge()`. The problem is that the retrieved data can be either:

* plain `array` — that’s not a problem :slightly_smiling_face:;
* or `Traversable` (specifically, it should be instance of `Zend\Session\Storage\StorageInterface`), but such type is not acceptable as `ArrayUtils::merge()` parameter :disappointed:.

The point of this PR is to convert such `Traversable` to a plain `array`, by calling `toArray()` (defined in the mentioned `StorageInterface`) on it. I’ve created also an alternative PR in `zend-stdlib` (zendframework/zend-stdlib#86) that’d let `ArrayUtils::merge()` accept `iterable` parameters.

gszy committed:
Don’t pass Traversable as ArrayUtils::merge parameter

If session data retrieved from $_SESSION (in SessionManager::start) was
instance of Traversable, it would be incorrectly passed to
ArrayUtils::merge, as that method accepts only plain arrays.

Ocramius commented:
Needs tests

gszy committed:
Test if SessionManager converts Traversable $_SESSION to array

gszy commented:
Would gscscnd/zend-session@9c5a1e75237a9997375cc9e9b821fe2ca8fe5c55 be enough or more tests are needed?

michalbundyra commented on src/SessionManager.php:
```diff
@@ -130,9 +130,13 @@ public function start($preserveStorage = false)
+            $oldSessionData = $oldSessionData->toArray();
+        }
+
+        if (! empty($oldSessionData) && is_array($oldSessionData)) {
```
We can have here just `if ($oldSessionData && is_array($oldSessionData)) {`

michalbundyra commented on src/SessionManager.php:
```diff
@@ -130,9 +130,13 @@ public function start($preserveStorage = false)
 
         session_start();
 
-        if ($oldSessionData instanceof \Traversable
```
Looks like this condition is removed. What then in case when `$oldSessionData instanceof \Traversable` (`=== true`) ?

gszy replied:
~~May I ask what _should_ be done in such case?~~ Is there a PHP (or ZF) function that converts `Traversable` to `array` or should I write something like:

```php
if ($oldSessionData instanceof \Traversable) {
    $oldSessionDataArray = [];

    foreach ($oldSessionData as $oldSessionDataKey => $oldSessionDataValue) {
        $oldSessionDataArray[$oldSessionDataKey] = $oldSessionDataValue;
    }

    $oldSessionData = $oldSessionDataArray;
}
```

Update: `iterator_to_array` would probably do the job.

michalbundyra replied:
See: http://php.net/manual/en/function.iterator-to-array.php

gszy replied:
Yeah, I’ve updated the patch.

michalbundyra replied:
One more thing to note: `iterator_to_array` is not recursive

@gscscnd Can you provide an example when `$_SESSION` is not an array?

gszy replied:
That is, before being passed to `ArrayUtils::merge`? I’ve encountered #10 when implementing “remember me” functionality — I had to retrieve `SessionManager` and I wasn’t able to do it until I’ve provided the following (minimal) configuration:

```php
return [
    'session_config' => [],
    'session_storage' => ['type' => Zend\Session\SessionStorage::class],
];
```

What happened is probably the same as described in #10. Yet another workaround: 8trust/zend-session@c5e523a4b661e3fb336bd009dc21f8edfe687270

michalbundyra replied:
@gscscnd thanks, now I see the issue, basically it's not possible to use other session storage than (session)array(storage), right?

gszy replied:
I haven’t tried all storage variants, but it seems `SessionManager` is not fully compatible with `SessionStorage` (and perhaps not only this), so either of them should be somehow fixed :confused:

Update: `SessionArrayStorage` seems to work for me. But I don’t think it’s an acceptable solution — while the docs say that `SessionStorage` may be incompatible with 3rd party libraries, it should be compatible with `zend-stdlib`, shouldn’t it?

michalbundyra commented on src/SessionManager.php:
```diff
@@ -130,9 +130,13 @@ public function start($preserveStorage = false)
-        if ($oldSessionData instanceof \Traversable
-            || (! empty($oldSessionData) && is_array($oldSessionData))
-        ) {
+        if ($oldSessionData instanceof Storage\StorageInterface) {
```
I think I don't understand something....
https://github.com/zendframework/zend-session/blob/7432b7dfde429a64e090c3cea7c8de99b1faeab5/src/Storage/StorageInterface.php#L23

Shouldn't we just have here: `if ($oldSessionData instanceof \Traversable) {` instead?

gszy replied:
`toArray` is defined in `StorageInterface` and not in the other interfaces.

michalbundyra replied:
Yes, see my above comments https://github.com/zendframework/zend-session/pull/93#discussion_r153446047

gszy commented:
Sorry, I didn’t notice this PR is aiming to do the same as #57

gszy committed:
Convert Traversable to array even if it’s not StorageInterface

gszy committed:
Better SessionManager::start() unit tests

$_SESSION is populated with StorageInterface (and then Traversable).
After start(), $_SESSION is checked whether it is an array and if still
contains initial data.

michalbundyra commented on src/SessionManager.php:
```diff
@@ -130,9 +130,15 @@ public function start($preserveStorage = false)
+        }
+
+        if (! empty($oldSessionData) && is_array($oldSessionData)) {
             $_SESSION = ArrayUtils::merge($oldSessionData, $_SESSION, true);
```
How it's going to work? As I can see `merge` has type hint on param 1 and 2 (`array`).
https://github.com/zendframework/zend-stdlib/blob/43d209995f054fa62557652192c4943625b4145a/src/ArrayUtils.php#L269
Your fix is in case when `$_SESSION` is NOT an array but above you pass (not an array) as the 2nd param?

gszy replied:
If I understand the code correctly, the new `$_SESSION` is passed to `merge()` after `session_start()` was called. Have a look at the [original code](https://github.com/zendframework/zend-session/blob/release-2.8.0/src/SessionManager.php#L126-L137) — it’s assumed (I guess so, because there’s no validation) that the new `$_SESSION` is something acceptable by `merge`. `session_start()` somehow modifies the `$_SESSION` superglobal.

I should move that `session_start()` call below my new code to make it easier to understand, shouldn’t I?

Update: moved.

gszy committed:
Moved $_SESSION conversion logic

Moved code that conditionally converts data fetched from $_SESSION to
plain array a few lines up for better readability.

weierophinney commented:
Thanks, @gscscnd; I have merged your changes along with #57 to provide a comprehensive fix.

gszy commented:
@weierophinney, I’m not entirely sure if https://github.com/zendframework/zend-session/blob/master/src/SessionManager.php#L142-L144 is completely correct:

* `toArray()` is defined in `StorageInterface`, not `Traversable` (or am I missing something?). Is it safe to assume that if `$_SESSION` content is `Traversable`, it is also `StorageInterface`? And I think that from static analysis tools point of view it would be necessary to compare against the latter interface. Finally, if it isn’t safe to assume that `$_SESSION` being instance of `Traversable` implies being instance of `StorageInterface`, there should be alternative conversion method provided, like https://github.com/zendframework/zend-session/blob/master/src/SessionManager.php#L135-L137.
* Now that there’s `use Traversable` on the top, `\Traversable` can be simplified to `Traversable` in https://github.com/zendframework/zend-session/blob/master/src/SessionManager.php#L135.

**Update**: #96 apparently aims to solve these problems.
